synapse/lib/layer.py

```diff
@@ -4093,6 +4094,103 @@ async def iterNodeDataKeys(self, buid):
             prop = self.getAbrvProp(abrv)
             yield prop[0]
 
+    async def iterLayerAddPerms(self):
+
+        # nodes & props
+        for byts, abrv in self.propabrv.slab.scanByFull(db=self.propabrv.name2abrv):
+            form, prop = s_msgpack.un(byts)
+            if form is None:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.byprop):
+                if prop:
+                    yield ('node', 'prop', 'set', f'{form}:{prop}')
+                else:
+                    yield ('node', 'add', form)
+
+        # tagprops
+        for byts, abrv in self.tagpropabrv.slab.scanByFull(db=self.tagpropabrv.name2abrv):
+            info = s_msgpack.un(byts)
+            if None in info or len(info) != 3:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.bytagprop):
+                yield ('node', 'tag', 'add', *info[1].split('.'))
+
+        # nodedata
+        for abrv in self.dataslab.scanKeys(db=self.dataname):
+            name, _ = self.getAbrvProp(abrv)
+            yield ('node', 'data', 'set', name)
+
+        # edges
+        for verb in self.layrslab.scanKeys(db=self.byverb):
+            yield ('node', 'edge', 'add', verb.decode())
+
+        # tags
+        # NB: tag perms should be yielded for every leaf on every node in the layer
+        tags = await s_spooled.Dict.anit(cell=self.core)
+
+        @s_cache.memoize()
+        def tagPartsByAbrv(abrv):
+            name = self.tagabrv.abrvToName(abrv)
+            return tuple(name.split('.'))
+
+        for lkey, buid in self.layrslab.scanByFull(db=self.bytag):
+            abrv = lkey[:8]
+            abrvs = list(tags.get(buid, []))
+            abrvs.append(abrv)
+            await tags.set(buid, abrvs)
+
+        # Perms we've already yielded
+        tagperms = await s_spooled.Set.anit(cell=self.core)
+
+        for buid, abrvs in tags.items():
+            seen = {}
+
+            if len(abrvs) == 1:
+                # Easy optimization: If there's only one tag abrv, then it's a
+                # leaf by default
+                key = tagPartsByAbrv(abrv)
+                if key not in tagperms:
+                    yield ('node', 'tag', 'add', *key)
+                    await tagperms.add(key)
+
+            else:
+                for abrv in abrvs:
+                    parts = tagPartsByAbrv(abrv)
+                    for idx in range(1, len(parts) + 1):
+                        key = tuple(parts[:idx])
+                        seen.setdefault(key, 0)
+                        seen[key] += 1
+
+                for key, count in seen.items():
+                    if count == 1 and key not in tagperms:
+                        yield ('node', 'tag', 'add', *key)
+                        await tagperms.add(key)
+
+    async def iterLayerDelPerms(self): # pragma: no cover
+        # TODO: Implement me for Syn3.x
+        pass
```

Comment:
One last nitpick, I don't think we'll need this since the two versions of the perms are "What do you need to create everything in the layer" (`iterLayerAddPerms`) and "What do you need to delete everything in the layer" (`iterLayerWipePerms`).

Reply:
For example a node with a tombstone in it does require a `del` perm but you're still "adding" that to the layer underneath when merging if that makes sense.